Enhance CalDateTime and DateTimeSerializer

axunonb opened the pull request:
Breaking Changes
* Removef the redundant `Calendar` argument to constructor overload of `CalDateTime`
* `CalDateTilme` only uses `DateTilmeKind.Utc` and `DateTimeKind.Unspecified`. `DateTimeKind.Local` has been dropped.
* Setting `null` for the timezone will apply the system`s local time with`AsDateTimeOffset`.
* The given timezone always takes precedence over any `DateTimeKind` of a `DateTime` value. This is also the case, if a UTC date and a timezone (not null) is used.
* `CalDateTime.TzId`: Will not read or write `CalDateTime.Parameters`. The functionality has been moved to `DateTimeSerializer.Deserialize(TextReader)
* `DateTimeSerializer.Deserialize(TextReader)` will now throw for unrepresentable date/time values.
* `CalDateTime.TzId` can be `null` or a valid timezone ID. Whitespace is not allowed.

CalDateTime
* Implement`DateOnly` and `TimeOnly` in order to return correct values for `CalDateTime.HasTime` and `CalDateTime.HasDate`, even if hour, minutes and seconds are zero.
* Updating `Value` or `TzId` will re-initialize the instance in the same way as with CTOR overloads.
* Refactored `ToString` overloads to convert the date and time to a string representation. Thanks to @nzbart for his initial PR.
* marked some properties as obsolete. 
* Updated related tests and methods to accommodate these changes.
* Enable NRT
* Complete xmldoc
* Increase code coverage to 94%

DateTimeSerializer
* Added validation in the `SerializeToString` method to ensure the TZID parameter is correctly handled based on whether the date-time is in UTC.
* Improved the Deserialize method to handle both full date-time and date-only strings, initializing the `DateOnly` and `TimeOnly` parts accordingly.
* Enable NRT
* Complete xmldoc
* Increase code coverage to 94%

Other
* Enable NRT for `IDateTime`
* Added new package reference `Portable.System.DateTimeOnly` for netstandard2.0 and netstandard2.1 for compatibility with net6.0 and later.

Resolves #630
Resolves #633
Resolves #635
Resolves #636
Resolves #637
Resolves #197

axunonb committed:
Enhance CalDateTime and DateTimeSerializer

* Introduced new methods and test cases in `CalDateTimeTests.cs` to ensure correct behavior of `CalDateTime` properties and methods.
* Updated various tests in `DeserializationTests.cs` and `RecurrenceTests.cs` to handle `CalDateTime` without time components and improve clarity.
* Refined `IsAllDay` property in `CalendarEvent.cs` and updated methods in `UniqueComponent.cs` and `VTimeZone.cs` to use `CalDateTime` with time components.
* Removed outdated comments and improved code formatting.
* Enabled nullable reference types and updated `IDateTime` interface.
* Added new methods and properties to `CalDateTime.cs` for better date and time management.
* Refactored methods in `Period`, `RecurrencePatternEvaluator`, and `RecurringEvaluator` classes to handle `HasTime` property correctly.
* Improved `DateTimeSerializer` class for better performance and handling of nullable types.
* Added XML documentation and marked obsolete methods.

Resolves #630
Resolves #633
Resolves #635
Resolves #636
Resolves #637

axunonb commented on Ical.Net/Evaluation/RecurrencePatternEvaluator.cs:
```diff
@@ -7,45 +7,6 @@
-    /// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
-    /// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
-    /// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
-    /// </summary>
```
Comparing our code base with the current iCal4j code its hard to say wether there are similarities left. The mention goes back to 2009. Even more important is, that their BSD 3-Clause License is not compatible to our MIT License.

axunonb commented on Ical.Net/DataTypes/CalDateTime.cs:
```diff
@@ -303,52 +303,12 @@ public override int GetHashCode()
-                    : Value.ToLocalTime().Date;
-            }
-        }
+        public DateTime AsSystemLocal => AsDateTimeOffset.LocalDateTime;
```
Use existing code from AsSystemLocal

axunonb commented on Ical.Net/DataTypes/CalDateTime.cs:
```diff
@@ -303,52 +303,12 @@ public override int GetHashCode()
-                return asUtc;
-            }
-        }
+        public DateTime AsUtc => AsDateTimeOffset.UtcDateTime;
```
Use existing code from AsSystemLocal

axunonb commented on Ical.Net/DataTypes/CalDateTime.cs:
```diff
@@ -533,7 +493,7 @@ public IDateTime ToTimeZone(string? tzId)
         public DateTimeOffset AsDateTimeOffset =>
             string.IsNullOrWhiteSpace(TzId)
-                ? new DateTimeOffset(AsSystemLocal)
+                ? new DateTimeOffset(Value)
```
Value contains the right `DateTimeKind` to create the `DatTimeOffset`

axunonb commented on Ical.Net/DataTypes/CalDateTime.cs:
```diff
@@ -169,146 +230,174 @@ public override int GetHashCode()
-            get => _hasTime;
-            set => _hasTime = value;
+            get => _timeOnly.HasValue;
+            set => _timeOnly = value ? TimeOnly.FromDateTime(_value) : null;
```
We can toggle on/off this way

axunonb commented on Ical.Net/DataTypes/CalDateTime.cs:
```diff
@@ -169,146 +230,174 @@ public override int GetHashCode()
-            get => _hasDate;
-            set => _hasDate = value;
+            get => _dateOnly.HasValue;
+            set => _dateOnly = value ? DateOnly.FromDateTime(_value) : null;
```
We toggle this, too, just to make IDateTime happy

axunonb commented on Ical.Net/DataTypes/CalDateTime.cs:
```diff
@@ -325,63 +414,65 @@ public string TzId
-                if (isEmpty)
+                if (string.IsNullOrWhiteSpace(value))
                 {
+                    Initialize(_value, _timeOnly.HasValue, value, Calendar);
```
Changing the TzId now initializes with immediate effect

axunonb commented on Ical.Net/DataTypes/Period.cs:
```diff
@@ -101,9 +104,7 @@ private void ExtrapolateTimes()
-            get => _startTime.HasTime
-                ? _startTime
-                : new CalDateTime(new DateTime(_startTime.Value.Year, _startTime.Value.Month, _startTime.Value.Day, 0, 0, 0), _startTime.TzId);
+            get => _startTime;
```
_startTime is ready to use by default

axunonb commented on Ical.Net/DataTypes/Period.cs:
```diff
@@ -137,7 +138,7 @@ public virtual TimeSpan Duration
             {
                 if (StartTime != null
                     && EndTime == null
-                    && StartTime.Value.TimeOfDay == TimeSpan.Zero)
```
must use HasTime

axunonb commented on Ical.Net/DataTypes/CalDateTime.cs:
```diff
@@ -169,146 +230,174 @@ public override int GetHashCode()
-                _asUtc = DateTime.MinValue;
-                _value = value;
+                // Initialize with the new value, keeping current 'HasTime' setting
+                Initialize(value, _timeOnly.HasValue, TzId, Calendar);
```
Setting the value initializes in the same as with using the CTOR overloads

minichma commented on Ical.Net/DataTypes/CalDateTime.cs:
```diff
@@ -1,130 +1,185 @@
 {
+    // The date and time parts that were used to initialize the instance
+    // or by the Value setter.
+    private DateTime _value;
```
I think, we shouldn't maintain both, the DateTime value *and* Date-/TimeOnly. I suggest to remove `_value` altogether.

axunonb replied:
Without the backing `value` we lose the `HasDate` and `HasTime` "toggle". The Date-/TimOnly fields are just the replacement for the backing `_hasDate` and `_hasTime` booleans. This implementation is mainly because `HasDate` and `HasTime` require a setter from the `IDateTime` interface.

minichma replied:
I thought it would be an option to calculate the DateTime value from the other two just in time, when needed.

axunonb replied:
I think I get your point: `HasDate` and `HasTime` wouldn't have backing fields, they would just become pure boolean getters/setters. Then `_value` can disappear. Yep, that's better.

axunonb replied:
Will be updated as proposed.

axunonb replied:
Done

minichma commented on Ical.Net/DataTypes/CalDateTime.cs:
```diff
@@ -1,130 +1,185 @@
-    private void Initialize(DateTime value, string tzId, Calendar cal)
-    {
-        if (!string.IsNullOrWhiteSpace(tzId) && !tzId.Equals("UTC", StringComparison.OrdinalIgnoreCase))
+        if ((tzId != null && !string.IsNullOrWhiteSpace(tzId) && !tzId.Equals("UTC", StringComparison.OrdinalIgnoreCase))
```
I know, it's legacy (so nothing to necessarily deal with in this PR), but I think that we shouldn't have special handling for white space, or empty strings. Either the value is set (!= null) or not (== null). Input sanitization is not something this class should be responsible for.

axunonb replied:
Will be updated as proposed.

axunonb replied:
Done

minichma commented on Ical.Net/DataTypes/CalDateTime.cs:
```diff
@@ -1,130 +1,185 @@
-            TzId = tzId;
+            _tzId = tzId;
+
+            initialValue = DateTime.SpecifyKind(dateTime, DateTimeKind.Local);
```
Also legacy, but IMHO important: I think we should avoid `DateTimeKind.Local` wherever we can, because dealing with the system's local time really is a special case. If no tzid is specified, this would usually mean that the instance represents floating time (which btw cannot be converted to a different time zone). Floating time should rather be represented by `DateTimeKind.Unspecified` than `.Local`.

axunonb replied:
Will be updated as proposed.

axunonb replied:
Done

axunonb committed:
Fix occasual failing unit test

Depending on the time offset from local time to UTC the tests failed.
Changed all date kinds to UTC.

axunonb committed:
Make AlmostZeroEpsilon private const

axunonb committed:
Replace DateTimeKind.Local with DateTimeKind.Unspecified

Affected classes:
* `CalDateTime`
* `DateUtil`
* `DateTimeSerializer`

axunonb committed:
SOC: Move getting "TZID" from `IParameterCollection` to `DateTimeSerializer`

`CalDateTime.TzId` does no more read or write `CalDateTime.Parameters`

axunonb committed:
Use backing fields `DateOnly` and `TimeOnly` to create the `CalDateTime.Value`

* Depending on `HasTime`, `Value` contains only the `DateTime.Date`, or the value including `DateTime.TimeOfDay`.
* The timezone dermines whether `DateTimeKind.Utc`or `DateTimeKind.Unspecified`will be used.

axunonb committed:
Remove the redundant `Calendar` argument to CTORs

minichma commented on Ical.Net/DataTypes/CalDateTime.cs:
```diff
@@ -174,271 +237,301 @@ public override int GetHashCode()
     {
-        var copy = left.Copy<IDateTime>();
+        var copy = left.Copy<CalDateTime>();
+        if (Math.Abs(right.TotalDays % 1) > AlmostZeroEpsilon)
```
Maybe `(right.Ticks % TimeSpan.TicksPerDay) != 0`? This way we could avoid dealing with FP precision issues.

axunonb replied:
excellent, done

minichma commented on Ical.Net/DataTypes/CalDateTime.cs:
```diff
@@ -174,271 +237,301 @@ public override int GetHashCode()
-                    ? Value
-                    : Value.Date;
-            }
+    public DateTime AsSystemLocal => AsDateTimeOffset.LocalDateTime;
```
I think we should remove `AsSystemLocal` altogether. `AsDateTimeOffset.LocalDateTime` can also be executed by the caller if needed.

axunonb replied:
So let's make a date/time without a timezone a "floating time". Then:
* `IsUtc` should return `false` - sounds reasonable
* `AsUtc` should throw, because a floating time (`TzId == null`) cannot be represented as UTC?
* `AsDateTimeOffset` should throw for the same reason
* and sure, `AsSystemLocal` must be removed, as it uses `TzId == null`
* add `IsFloating`?

axunonb replied:
Must be done separately, because `AsSystemLocal` is a member of `IDateTime` - see new task in #646

minichma commented on Ical.Net/DataTypes/CalDateTime.cs:
```diff
@@ -1,130 +1,188 @@
 {
-    public static CalDateTime Now => new CalDateTime(DateTime.Now);
+    // The date part that is used to return the Value property.
+    private DateOnly _dateOnly;
```
Shouldn't we define `_dateOnly` and `_timeOnly` as nullable? Then we wouldn't have to explicitly store `HasTime` and it would get become natural as

```cs
    public bool HasTime {
        get => this._timeOnly != null;
        set => this._timeOnly = (value ? (this._timeOnly ?? new TimeOnly()) : null);
    }
```

We would remove additional redundancy.

axunonb replied:
The setter would then overwrite the original value that is needed in the current v4 implementation.
I did not yet investigate implications, and thus made it compatible.

In the end, `HasTime` and `HasDate`, `TzId` and `Value` should just have getters. This requires to change (or even remove entirely) `IDateTime`. That gets to huge for this PR.

minichma replied:
The setter would only overwrite, if the value is changed true->false->true, which I would consider legit in that case. But I fully agree that HasTime/HasDate setters should be removed altogether and that this should be covered in different PR. Also the additional considerations regarding local time should be handled in a different PR, so this one doesn't get bloated too much.

axunonb replied:
Created issue  #646 for this task - will not include it into this PR

minichma commented on Ical.Net/Utility/DateUtil.cs:
```diff
@@ -22,7 +22,7 @@ public static IDateTime EndOfDay(IDateTime dt)
 
     public static DateTime GetSimpleDateTimeData(IDateTime dt)
-        => DateTime.SpecifyKind(dt.Value, dt.IsUtc ? DateTimeKind.Utc : DateTimeKind.Local);
+        => DateTime.SpecifyKind(dt.Value, dt.IsUtc ? DateTimeKind.Utc : DateTimeKind.Unspecified);
```
According to the improved CalDateTime, this would be as simple as `=> dt.Value`.

axunonb replied:
right - done

minichma commented on Ical.Net/Utility/DateUtil.cs:
I think, there are a few more places to deal with local date in this class. 

* The behavior of `SimpleDateTimeToMatch` is not easy to understand. Within the sln its only used for testing, so I'd suggest to move it to a test project, where the specific behavior might be helpful.
* `MatchTimeZone` is only used in `RecurrencePatternEvaluator`. I'm pretty sure, that the way it deals with floating time is not the most natural one either (see #197). Probably we want to deal with that in a separate PR. Maybe move this method into `RecurrencePatternEvaluator` for now and/or add a comment that we should reconsider/fix how it deals with local time.

axunonb replied:
Done:
* Move `DateUtil.SimpleDateTimeToMatch` as private to `RecurrenceTests`
* Move `DateUtil.MatchTimeZone` as private  to `RecurrencePatternEvaluator`

minichma replied:
`DateUtil` is internal, so this shouldn't be breaking.

minichma commented on Ical.Net/DataTypes/CalDateTime.cs:
```diff
@@ -1,130 +1,188 @@
+    /// <summary>
+    /// Gets the current date/time in the local timezone.
+    /// </summary>
+    public static CalDateTime Now => new CalDateTime(DateTime.Now, null, true);
```
If we want to offer a `Now`, then this is one of the very few places where referring to the system's local time would make sense.

minichma commented on Ical.Net/DataTypes/CalDateTime.cs:
```diff
@@ -1,130 +1,188 @@
+    private TimeOnly _timeOnly;
 
-    public static CalDateTime Today => new CalDateTime(DateTime.Today);
+    private const double AlmostZeroEpsilon = 1e-10;
```
See my comment on `operator -`. I think, we can having this kind of epsilon.

minichma commented on Ical.Net/DataTypes/CalDateTime.cs:
```diff
@@ -1,130 +1,188 @@
+    /// the system's local timezone.
+    /// </param>
+    /// <param name="hasTime">Set to <see langword="true"/> (default), if the <see cref="DateTime.TimeOfDay"/> must be included.</param>
+    public CalDateTime(DateTime value, string? tzId, bool hasTime = true)
```
Having both, the `Kind` in the `DateTime` value, as well as the `tzId`, causes redundancy, that's not easy to understand. What, if we remove this constructor altogether? The user could easily call `new CalDateTime(value).ToTimeZone(tzId)`, with way less ambiguity.

axunonb replied:
This will be back on the agenda when we implement floating date/time. See #647

minichma commented on Ical.Net/DataTypes/CalDateTime.cs:
```diff
@@ -174,271 +237,301 @@ public override int GetHashCode()
     public DateTime Date => Value.Date;
 
+    /// <inheritdoc cref="DateTime.TimeOfDay"/>
     public TimeSpan TimeOfDay => Value.TimeOfDay;
```
Isn't this obsolete, now that we have the `TimeOnlyValue`?

axunonb replied:
see task in https://github.com/ical-org/ical.net/issues/646

minichma commented on Ical.Net/DataTypes/CalDateTime.cs:
```diff
@@ -174,271 +237,301 @@ public override int GetHashCode()
     public int DayOfYear => Value.DayOfYear;
 
+    /// <inheritdoc cref="DateTime.Date"/>
     public DateTime Date => Value.Date;
```
Obsolete with `DateOnlyValue`?

axunonb replied:
see task in #646

minichma commented on Ical.Net/DataTypes/CalDateTime.cs:
```diff
@@ -174,271 +237,301 @@ public override int GetHashCode()
+    /// <summary>
+    /// Gets the underlying <see cref="DateOnlyValue"/> of <see cref="Value"/>.
+    /// </summary>
+    public DateOnly? DateOnlyValue => HasDate ? _dateOnly : null;
```
Maybe just name it `Date`, in replacement of the existing, now obsolete `Date`?

axunonb replied:
see task in https://github.com/ical-org/ical.net/issues/646

minichma commented on Ical.Net/DataTypes/CalDateTime.cs:
```diff
@@ -174,271 +237,301 @@ public override int GetHashCode()
+    /// <summary>
+    /// Gets the underlying <see cref="TimeOnlyValue"/> of <see cref="Value"/>.
+    /// </summary>
+    public TimeOnly? TimeOnlyValue => HasTime ? _timeOnly : null;
```
Just `Time` or `TimeOfDay`?

axunonb replied:
see task in https://github.com/ical-org/ical.net/issues/646

minichma reviewed (commented):
I love the simplification that comes with removing the ambiguous `DateTime` member field and with removing `DateTimeKind.Local`! Didn't do a full review, only `CalDateTime` for now, but I think that's the core part anyways. Left a few comments with my thoughts, maybe not all relevant for this PR but for following ones.

axunonb committed:
Merge branch 'version/v5' into pr/caldatetime2

axunonb committed:
Implement review comments

* `if (Math.Abs(right.TotalDays % 1) > AlmostZeroEpsilon)` is now `if ((right.Ticks % TimeSpan.TicksPerDay) != 0)`
* `DateUtil.DateTime GetSimpleDateTimeData(IDateTime dt)` returns `dt.Value`
* Move internal `DateUtil.SimpleDateTimeToMatch` as private to `RecurrenceTests`
* Move internal `DateUtil.MatchTimeZone` as private  to `RecurrencePatternEvaluator`
* Create tasks in #646 and #647 for new PRs

axunonb committed:
Merge branch 'version/v5' into pr/caldatetime2

axunonb committed:
Remove xmldoc for non-existing parameter "kind"

Also refactor for SonarCloud complaints
